contracts/mixnet/src/transactions.rs

```diff
@@ -3,15 +3,63 @@ use crate::error::ContractError;
 use crate::helpers::{calculate_epoch_reward_rate, scale_reward_by_uptime};
 use crate::state::StateParams;
 use crate::storage::{
-    config, config_read, decrement_layer_count, gateway_delegations, gateways, gateways_owners,
-    gateways_owners_read, gateways_read, increase_gateway_bond, increase_gateway_delegated_stakes,
-    increase_mix_delegated_stakes, increase_mixnode_bond, increment_layer_count, mix_delegations,
-    mixnodes, mixnodes_owners, mixnodes_owners_read, mixnodes_read, read_gateway_epoch_reward_rate,
-    read_mixnode_epoch_reward_rate, read_state_params, Layer,
+    config, config_read, decrement_layer_count, gateway_delegations, gateway_delegations_read,
+    gateways, gateways_owners, gateways_owners_read, gateways_read,
+    increase_gateway_delegated_stakes, increase_mix_delegated_stakes, increment_layer_count,
+    mix_delegations, mix_delegations_read, mixnodes, mixnodes_owners, mixnodes_owners_read,
+    mixnodes_read, read_gateway_epoch_reward_rate, read_mixnode_epoch_reward_rate,
+    read_state_params, Layer,
 };
-use cosmwasm_std::{attr, coins, BankMsg, Coin, Decimal, DepsMut, MessageInfo, Response, Uint128};
+use cosmwasm_std::{
+    attr, coins, BankMsg, Coin, Decimal, DepsMut, MessageInfo, Order, Response, StdResult, Uint128,
+};
+use cosmwasm_storage::ReadonlyBucket;
 use mixnet_contract::{Gateway, GatewayBond, IdentityKey, MixNode, MixNodeBond};
 
+const OLD_DELEGATIONS_CHUNK_SIZE: usize = 500;
+
+// Looks for the total amount of delegations towards a particular node.
+// This function is used only in very specific circumstances:
+// 1. The mixnode/gateway bonds
+// 2. Some addresses start to delegate to the node
+// 3. The node unbonds
+// 4. Some of the addresses that delegated in the past have not removed the delegation yet
+// 5. The node rebonds with the same identity
+fn find_old_delegations(delegations_bucket: ReadonlyBucket<Uint128>) -> StdResult<Coin> {
+    // I think it's incredibly unlikely to ever read more than that
+    // but in case we do, we should guard ourselves against possible
+    // out of memory errors (wasm contracts can only allocate at most 2MB
+    // of RAM, so we don't want to box the entire iterator)
+    let mut total_delegation = Coin::new(0, DENOM);
+    let mut start = None;
+    loop {
+        let iterator = delegations_bucket
```

Comment:
Superficially worried about looping in a loop, but not too much

Reply:
Yeah, ideally I would have avoided it, but this particular iterator doesn't have `size_hint()` properly defined hence I do not know its size beforehand. Also, I couldn't iterate everything at once as when the iterator is created, everything is being allocated and in the worst case scenario we would have gone over the 2MB limit of the wasmVM. 

I hope that the test I've written for it would have caught any infinite loop issues